### PR TITLE
SDK-681 Changes SDK Side due to fixes on Circuit V2

### PR DESCRIPTION
**Pull request by sgonta**
https://horizenlabs.atlassian.net/browse/SDK-681

### CODE REVIEW

**Review by i-Alex — CHANGES_REQUESTED** [review]
please, undo the commit and remove all `libevm` files from it, then force push again to remote

File: sdk/src/main/scala/com/horizen/api/http/SidechainSubmitterApiRoute.scala [hunk 1]
```diff
@@ -82,17 +80,17 @@ case class SidechainSubmitterApiRoute(override val settings: RESTApiSettings, ce
...
-          RespHashSchnorrPublicKey(
-            BytesUtils.toHexString(schnorrPublicKey.getHash)
-          )
-        )
-      } catch {
-        case e: Throwable => SidechainApiError(e)
+  def getKeyRotationMessageToSign: Route = (post & path("getKeyRotationMessageToSign")) {
+    entity(as[ReqGetKeyRotationMessageToSign]) { body =>
+      Try {
+        Await.result(certSubmitterRef ? GetKeyRotationMessageToSign(body.schnorrPublicKey, body.keyType,
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/api/http/SidechainSubmitterApiRoute.scala:86]: Let's simplify code here.
We don't need to call CertificateSubmitter to get key message to sign. Because Submitter actually doesn't care about it.
Instead you may directly use the following code here
```
if(keyType == 0) {
        CryptoLibProvider.thresholdSignatureCircuitWithKeyRotation.getMsgToSignForSigningKeyUpdate(SchnorrPublicKey.deserialize(schnorrPublicKey.getBytes), withdrawalEpoch,  params.sidechainId)
      } else {
        CryptoLibProvider.thresholdSignatureCircuitWithKeyRotation.getMsgToSignForMasterKeyUpdate(SchnorrPublicKey.deserialize(schnorrPublicKey.getBytes), withdrawalEpoch,  params.sidechainId)
      }
```

File: sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala [hunk 2]
```diff
@@ -433,6 +435,16 @@ class CertificateSubmitter[T <: CertificateData](settings: SidechainSettings,
     }
   }
 
+  private def keyRotationMessageToSign: Receive = {
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala:438]: considering my comment in SidechainSubmitterApiRoute class, revert the changes in this file.
Same for CertificateSubmitterTest file.

File: sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala [hunk 3]
```diff
@@ -117,7 +118,8 @@ class CertificateSubmitter[T <: CertificateData](settings: SidechainSettings,
       getSignaturesStatus orElse
       submitterStatus orElse
       signerStatus orElse
-      reportStrangeInput
+      reportStrangeInput orElse
+      keyRotationMessageToSign
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala:122]: just a note for the future `reportStrangeInput`  must be the last case-processor in the chain

File: sdk/src/main/scala/com/horizen/api/http/SidechainSubmitterApiRoute.scala [hunk 4]
```diff
@@ -82,17 +80,17 @@ case class SidechainSubmitterApiRoute(override val settings: RESTApiSettings, ce
     ApiResponseUtil.toResponse(RespSubmitterOk)
   }
 
-  def getSchnorrPublicKeyHash: Route = (post & path("getSchnorrPublicKeyHash")) {
-    entity(as[ReqGetSchnorrPublicKeyHash]) { body =>
-      try {
-        val schnorrPublicKey: SchnorrProposition = new SchnorrProposition(BytesUtils.fromHexString(body.schnorrPublicKey))
-        ApiResponseUtil.toResponse(
-          RespHashSchnorrPublicKey(
-            BytesUtils.toHexString(schnorrPublicKey.getHash)
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/api/http/SidechainSubmitterApiRoute.scala:91]: Changes in the SidechainState are missing.
The code in the API was updated: instead of `SchnorrProposition.getHash` we now use `CryptoLibProvider.thresholdSignatureCircuitWithKeyRotation.getMsgToSignForSigningKeyUpdate`.
But in the State during verification we still use the old `SchnorrProposition.getHash`.
Check and update the following methods:
1) SidechainState.validateWithWithdrawalEpoch
2) SidechainStateState.getKeyRotationTransaction

Moreover, let's remove SchnorrProposition.getHash method at all, it will clearly show now and during Account model integration all places where we need to change the code.

File: qa/sc_cert_key_rotation_across_epoch.py [hunk 5]
```diff
@@ -17,7 +17,7 @@
 from httpCalls.submitter.getCertifiersKeys import http_get_certifiers_keys
 from httpCalls.submitter.getKeyRotationProof import http_get_key_rotation_proof
 from SidechainTestFramework.secure_enclave_http_api_server import SecureEnclaveApiServer
-from httpCalls.submitter.getSchnorrPublicKeyHash import http_get_schnorr_public_key_hash
+from httpCalls.submitter.getKeyRotationMessageToSign import http_get_key_rotation_message_to_sign
```
Comment (i-Alex) [inline qa/sc_cert_key_rotation_across_epoch.py:20]: We need to extend the test with one more case
Look at the description:
`- Create a CertificateKeyRotationTransaction (T3) to change the signer key 0 (SK0 -> SK3)`
Similar to that we need to create Tx T4 with valid key transition, but invalid message to sign, because of the "old" withdrawal epoch number. Same as with T3, we need to check that for the first block of the new epoch - tx is invalid.

File: qa/sc_cert_key_rotation.py [hunk 6]
```diff
@@ -17,7 +17,7 @@
 from SidechainTestFramework.secure_enclave_http_api_server import SecureEnclaveApiServer
 from httpCalls.submitter.getCertifiersKeys import http_get_certifiers_keys
 from httpCalls.submitter.getKeyRotationProof import http_get_key_rotation_proof
-from httpCalls.submitter.getSchnorrPublicKeyHash import http_get_schnorr_public_key_hash
+from httpCalls.submitter.getKeyRotationMessageToSign import http_get_key_rotation_message_to_sign
```
Comment (i-Alex) [inline qa/sc_cert_key_rotation.py:20]: add negative case, with invalid message to sign because of:
1) wrong key type
2) wrong epoch number

**Review by i-Alex — CHANGES_REQUESTED** [review]

File: sdk/src/main/resources/api/sidechainApi.yaml [hunk 7]
```diff
@@ -1894,24 +1894,36 @@ paths:
...
                   type: string
                   description: public key of certificate signer
+                keyType:
+                  type: integer
+                  format: int32
+                  minimum: 0
+                  maximum: 1
+                  description: Key type - 0 for signers key, 1 for masters key
+                withdrawalEpoch:
+                  type: integer
```
Comment (MarcoOl94) [inline sdk/src/main/resources/api/sidechainApi.yaml:1924]: we can add minimum: 0

File: sdk/src/main/scala/com/horizen/api/http/SidechainSubmitterApiRoute.scala (hunk 4 above)
Comment (MarcoOl94) [reply sdk/src/main/scala/com/horizen/api/http/SidechainSubmitterApiRoute.scala:91]: Update also the **SidechainStateTest** accordingly (e.g. _getKeyRotationTransaction()_, _keyRotationTransactionTest()_

**Review by MarcoOl94 — CHANGES_REQUESTED** [review]

File: qa/httpCalls/submitter/getKeyRotationMessageToSign.py [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import json
+
+
+# execute a submitter/getKeyRotationMessageToSign call
+def http_get_key_rotation_message_to_sign(sidechain_node, pub_key_to_sign, key_type, withdrawal_epoch):
```
[the remark this answers is not in the document]
Comment (sgonta) [reply qa/httpCalls/submitter/getKeyRotationMessageToSign.py]: Done

File: qa/sc_cert_key_rotation.py [hunk 9]
```diff
@@ -300,7 +302,7 @@ def run_test(self):
         # Change again the same signature key
         new_signing_key_2 = generate_cert_signer_secrets("random_seed2", 1)[0]
         new_public_key_2 = new_signing_key_2.publicKey
-        new_public_key_hash_2 = http_get_schnorr_public_key_hash(sc_node, new_public_key_2)["schnorrPublicKeyHash"]
+        new_public_key_hash_2 = http_get_key_rotation_message_to_sign(sc_node, new_public_key_2, key_type, withdrawal_epoch)["keyRotationMessageToSign"]
```
[the remark this answers is not in the document]
Comment (sgonta) [reply qa/sc_cert_key_rotation.py:304]: Done

File: sdk/src/main/scala/com/horizen/api/http/SidechainSubmitterApiRoute.scala (hunk 1 above)
Comment (sgonta) [reply sdk/src/main/scala/com/horizen/api/http/SidechainSubmitterApiRoute.scala]: Done

File: sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala (hunk 2 above)
Comment (sgonta) [reply sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala]: Done

File: qa/sc_cert_key_rotation.py [hunk 10]
```diff
@@ -503,11 +505,11 @@ def run_test(self):
         for i in range(self.cert_max_keys):
             new_signing_key = generate_cert_signer_secrets("random_seed5", 1)[0]
             new_signing_keys += [new_signing_key]
-            new_signing_key_hash = http_get_schnorr_public_key_hash(sc_node, new_signing_key.publicKey)["schnorrPublicKeyHash"]
+            new_signing_key_hash = http_get_key_rotation_message_to_sign(sc_node, new_signing_key.publicKey, key_type, withdrawal_epoch)["keyRotationMessageToSign"]
```
[the remark this answers is not in the document]
Comment (sgonta) [reply qa/sc_cert_key_rotation.py]: Done

File: qa/sc_cert_key_rotation.py [hunk 11]
```diff
@@ -163,7 +163,9 @@ def run_test(self):
         # Try to change the signing key 0
         new_signing_key = generate_cert_signer_secrets("random_seed", 1)[0]
         new_public_key = new_signing_key.publicKey
-        new_public_key_hash = http_get_schnorr_public_key_hash(sc_node, new_public_key)["schnorrPublicKeyHash"]
+        key_type = 0
+        withdrawal_epoch = self.sc_withdrawal_epoch_length
```
[the remark this answers is not in the document]
Comment (sgonta) [reply qa/sc_cert_key_rotation.py:167]: ok

File: sdk/src/main/scala/com/horizen/api/http/SidechainSubmitterApiRoute.scala (hunk 4 above)
Comment (sgonta) [reply sdk/src/main/scala/com/horizen/api/http/SidechainSubmitterApiRoute.scala:91]: I updated SidechainStateTest, but there are a lot of things to change there in order to make it work. Could you please checkout the branch SDK-681, pull the latest changes, run SidechainStateTest and see what can we modify?

File: qa/httpCalls/submitter/getKeyRotationMessageToSign.py [hunk 12]
```diff
@@ -2,12 +2,21 @@
...
+    return response["result"]
+
+
+def http_get_key_rotation_message_to_sign_for_signing_key(sidechain_node, pub_key_to_sign, withdrawal_epoch):
+    j = {
+        "schnorrPublicKey": pub_key_to_sign,
+        "withdrawalEpoch": withdrawal_epoch
+    }
+    request = json.dumps(j)
+    response = sidechain_node.submitter_getKeyRotationMessageToSignForMasterKey(request)
```
Comment (i-Alex) [inline qa/httpCalls/submitter/getKeyRotationMessageToSign.py:21]: the method name and the API endpoint name are different.
http_get_key_rotation_message_to_sign_for_**signing_key**
submitter_getKeyRotationMessageToSignFor**MasterKey**

Same for other method above.

File: qa/sc_cert_key_rotation.py [hunk 13]
```diff
@@ -302,7 +302,7 @@ def run_test(self):
         # Change again the same signature key
         new_signing_key_2 = generate_cert_signer_secrets("random_seed2", 1)[0]
         new_public_key_2 = new_signing_key_2.publicKey
-        new_public_key_hash_2 = http_get_key_rotation_message_to_sign(sc_node, new_public_key_2, key_type, withdrawal_epoch)["keyRotationMessageToSign"]
+        new_public_key_hash_2 = http_get_key_rotation_message_to_sign_for_signing_key(sc_node, new_public_key_2, withdrawal_epoch)["keyRotationMessageToSign"]
```
Comment (i-Alex) [inline qa/sc_cert_key_rotation.py:305]: one more place left with outdated name `new_public_key_hash_2`

File: sdk/src/main/java/com/horizen/cryptolibprovider/implementations/ThresholdSignatureCircuitWithKeyRotationImplZendoo.java [hunk 14]
```diff
@@ -293,9 +293,10 @@ public byte[] getSchnorrKeysHash(SchnorrKeysSignatures schnorrKeysSignatures) {
     }
 
     @Override
-    public byte[] getMsgToSignForSigningKeyUpdate(SchnorrPublicKey newSigningKey, int epochNumber, byte[] sidechainId) {
+    public byte[] getMsgToSignForSigningKeyUpdate(byte[] newSigningKeyBytes, int epochNumber, byte[] sidechainId) {
         byte[] messageAsBytes;
         FieldElement sidechainIdFe = FieldElement.deserialize(sidechainId);
+        SchnorrPublicKey newSigningKey = SchnorrPublicKey.deserialize(newSigningKeyBytes);
```
Comment (i-Alex) [inline sdk/src/main/java/com/horizen/cryptolibprovider/implementations/ThresholdSignatureCircuitWithKeyRotationImplZendoo.java:299]: `newSigningKey.freePublicKey();` is missing below

File: sdk/src/main/scala/com/horizen/SidechainState.scala [hunk 15]
```diff
@@ -376,7 +376,8 @@ class SidechainState private[horizen] (stateStorage: SidechainStateStorage,
       val keyRotationProof = keyRotationTransaction.getKeyRotationProof
       val oldCertifiersKeys = certifiersKeys(withdrawalEpoch - 1).get
 
-      val messageToSign = keyRotationProof.newKey.getHash
+      val messageToSign = CryptoLibProvider.thresholdSignatureCircuitWithKeyRotation
+        .getMsgToSignForSigningKeyUpdate(keyRotationProof.newKey.pubKeyBytes(), withdrawalEpoch, params.sidechainId)
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/SidechainState.scala:380]: here we should match `keyRotationProof.keyType`
case signing -> getMsgToSignForSigningKeyUpdate
case master -> getMsgToSignForMasterKeyUpdate

File: sdk/src/test/scala/com/horizen/SidechainStateTest.scala [hunk 16]
```diff
@@ -119,7 +120,8 @@ class SidechainStateTest
 
   def getKeyRotationTransaction(boxesWithSecretToOpen: (ZenBox,PrivateKey25519), typeOfKey: KeyRotationProofTypes.KeyRotationProofType, keyIndex: Int, newKeySecret: SchnorrSecret, oldSigningKeySecret: SchnorrSecret, oldMasterKeySecret: SchnorrSecret, wrongNewKey: Boolean = false): CertificateKeyRotationTransaction = {
     val from: JPair[ZenBox,PrivateKey25519] =  new JPair[ZenBox,PrivateKey25519](boxesWithSecretToOpen._1, boxesWithSecretToOpen._2)
-    val messageToSign = newKeySecret.publicImage().getHash
+    val messageToSign = CryptoLibProvider.thresholdSignatureCircuitWithKeyRotation
+      .getMsgToSignForSigningKeyUpdate(newKeySecret.publicImage().pubKeyBytes(), 0, params.sidechainId)
```
Comment (i-Alex) [inline sdk/src/test/scala/com/horizen/SidechainStateTest.scala:124]: 1) take into account `typeOfKey` same as int the State. Use proper method getMsgToSignForSigningKeyUpdate or getMsgToSignForMasterKeyUpdate same as in the State.
2) add two more arguments to the `getKeyRotationTransaction` method:
* `withdrawalEpochNumber`. Use it here. The passed value must be equal to the one one in the test. Likely, now it is `0`.
* `params: NetworkParams` - at the moment you use `params` field of the test class, but instead you should consider the `mockedParams` local variable of the exact test method, where you need to define `sidechainId` that is used both here and in the `state.validate` later to avoid the nullpointerexception error in the test

**Review by i-Alex — CHANGES_REQUESTED** [review]

File: qa/sc_cert_key_rotation.py (hunk 13 above)
Comment (sgonta) [reply qa/sc_cert_key_rotation.py]: Done

File: sdk/src/main/scala/com/horizen/SidechainState.scala (hunk 15 above)
Comment (sgonta) [reply sdk/src/main/scala/com/horizen/SidechainState.scala:380]: Done

File: qa/sc_cert_key_rotation.py (hunk 6 above)
Comment (sgonta) [reply qa/sc_cert_key_rotation.py]: Wrong key type I added, wrong epoch number is impossible to check, because there's no response 
image

Comment (sgonta) [reply qa/sc_cert_key_rotation.py]: Wrong key type I added, wrong epoch number is impossible to check, because there's no response 
image

Comment (sgonta) [reply qa/sc_cert_key_rotation.py]: Actually key type is also checked on request level
image

**Review by i-Alex — APPROVED** [review]

**Comment by sgonta** [comment]
image

**Review by MarcoOl94 — APPROVED** [review]

**Comment by sgonta** [comment]
@i-Alex  Thank you for merging!
